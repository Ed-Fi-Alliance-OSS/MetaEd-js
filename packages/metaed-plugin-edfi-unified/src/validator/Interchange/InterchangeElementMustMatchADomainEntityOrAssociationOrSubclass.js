// @flow
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { failInterchangeItemNotMatchingBaseClassProperty } from '../ValidatorShared/FailInterchangeItemNotMatchingBaseClassProperty';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach(namespace => {
    failInterchangeItemNotMatchingBaseClassProperty(
      'InterchangeElementMustMatchADomainEntityOrAssociationOrSubclass',
      namespace.entity,
      'elements',
      'Interchange element',
      failures,
    );
  });
  return failures;
}
