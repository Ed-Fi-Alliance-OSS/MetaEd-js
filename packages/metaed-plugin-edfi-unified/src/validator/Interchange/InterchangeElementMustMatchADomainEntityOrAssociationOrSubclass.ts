import { MetaEdEnvironment, ValidationFailure, Interchange, InterchangeItem, ModelType } from 'metaed-core';
import { getAllEntitiesOfType, getEntityFromNamespaceChain } from 'metaed-core';

const validTypes: ModelType[] = ['association', 'associationSubclass', 'descriptor', 'domainEntity', 'domainEntitySubclass'];

const validTypeNames: string = [
  'Abstract Entity',
  'Association',
  'Association Subclass',
  'Descriptor',
  'Domain Entity',
  'Domain Entity Subclass',
].join(', ');

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  (getAllEntitiesOfType(metaEd, 'interchange') as Interchange[]).forEach((interchange: Interchange) => {
    if (interchange.elements.length === 0) return;
    interchange.elements.forEach((item: InterchangeItem) => {
      const foundEntity = getEntityFromNamespaceChain(
        item.metaEdName,
        item.referencedNamespaceName,
        interchange.namespace,
        ...validTypes,
      );
      if (foundEntity != null) return;
      failures.push({
        validatorName: 'InterchangeElementMustMatchADomainEntityOrAssociationOrSubclass',
        category: 'error',
        message: `Interchange element ${item.metaEdName} does not match any declared ${validTypeNames}`,
        sourceMap: item.sourceMap.metaEdName,
        fileMap: null,
      });
    });
  });
  return failures;
}
