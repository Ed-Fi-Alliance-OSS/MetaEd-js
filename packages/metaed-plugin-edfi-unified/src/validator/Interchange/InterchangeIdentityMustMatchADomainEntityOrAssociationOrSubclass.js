// @flow
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { failInterchangeItemNotMatchingBaseClassProperty } from '../ValidatorShared/FailInterchangeItemNotMatchingBaseClassProperty';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach(namespace => {
    failInterchangeItemNotMatchingBaseClassProperty(
      'InterchangeIdentityMustMatchADomainEntityOrAssociationOrSubclass',
      namespace.entity,
      'identityTemplates',
      'Interchange identity template',
      failures,
    );
  });
  return failures;
}
