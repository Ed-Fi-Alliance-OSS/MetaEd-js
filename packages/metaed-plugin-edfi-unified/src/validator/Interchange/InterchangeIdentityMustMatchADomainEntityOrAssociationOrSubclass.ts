import { MetaEdEnvironment, ValidationFailure, Interchange, InterchangeItem, ModelType } from 'metaed-core';
import { getAllEntitiesOfType, getEntityForNamespaces } from 'metaed-core';

const validTypes: ModelType[] = ['association', 'associationSubclass', 'descriptor', 'domainEntity', 'domainEntitySubclass'];

const validTypeNames: string = [
  'Abstract Entity',
  'Association',
  'Association Subclass',
  'Descriptor',
  'Domain Entity',
  'Domain Entity Subclass',
].join(', ');

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  (getAllEntitiesOfType(metaEd, 'interchange') as Array<Interchange>).forEach((interchange: Interchange) => {
    if (interchange.identityTemplates.length === 0) return;
    interchange.identityTemplates.forEach((item: InterchangeItem) => {
      const foundEntity = getEntityForNamespaces(
        item.metaEdName,
        [interchange.namespace, ...interchange.namespace.dependencies],
        ...validTypes,
      );
      if (foundEntity != null) return;
      failures.push({
        validatorName: 'InterchangeIdentityMustMatchADomainEntityOrAssociationOrSubclass',
        category: 'error',
        message: `Interchange identity template ${item.metaEdName} does not match any declared ${validTypeNames}`,
        sourceMap: item.sourceMap.metaEdName,
        fileMap: null,
      });
    });
  });
  return failures;
}
