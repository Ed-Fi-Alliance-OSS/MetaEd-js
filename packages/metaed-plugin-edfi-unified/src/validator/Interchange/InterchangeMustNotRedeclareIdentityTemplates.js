// @flow
import type { MetaEdEnvironment } from '../../../../../packages/metaed-core/src/MetaEdEnvironment';
import type { ValidationFailure } from '../../../../../packages/metaed-core/src/validator/ValidationFailure';
import { failInterchangeItemRedeclarations } from '../ValidatorShared/FailInterchangeItemRedeclarations';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.interchange.forEach(interchange => {
    if (interchange.identityTemplates.length === 0) return;
    failInterchangeItemRedeclarations('InterchangeMustNotRedeclareIdentityName', 'identity template', interchange, interchange.identityTemplates, failures);
  });
  return failures;
}
