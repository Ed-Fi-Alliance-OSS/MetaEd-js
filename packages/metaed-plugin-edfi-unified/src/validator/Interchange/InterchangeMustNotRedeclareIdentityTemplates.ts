import { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { failInterchangeItemRedeclarations } from '../ValidatorShared/FailInterchangeItemRedeclarations';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.namespace.forEach(namespace => {
    namespace.entity.interchange.forEach(interchange => {
      if (interchange.identityTemplates.length === 0) return;
      failInterchangeItemRedeclarations(
        'InterchangeMustNotRedeclareIdentityName',
        'identity template',
        interchange,
        interchange.identityTemplates,
        failures,
      );
    });
  });
  return failures;
}
