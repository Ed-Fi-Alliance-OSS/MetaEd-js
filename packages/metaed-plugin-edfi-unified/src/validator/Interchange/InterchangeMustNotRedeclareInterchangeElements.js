// @flow
import type { MetaEdEnvironment } from '../../../../../packages/metaed-core/src/MetaEdEnvironment';
import type { ValidationFailure } from '../../../../../packages/metaed-core/src/validator/ValidationFailure';
import { failInterchangeItemRedeclarations } from '../ValidatorShared/FailInterchangeItemRedeclarations';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.interchange.forEach(interchange => {
    if (interchange.elements.length === 0) return;
    failInterchangeItemRedeclarations('InterchangeMustNotRedeclareInterchangeElements', 'interchange element', interchange, interchange.elements, failures);
  });
  return failures;
}
