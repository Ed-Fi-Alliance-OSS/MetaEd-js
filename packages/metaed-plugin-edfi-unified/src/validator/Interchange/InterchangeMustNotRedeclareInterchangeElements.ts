import { MetaEdEnvironment, ValidationFailure } from '@edfi/metaed-core';
import { failInterchangeItemRedeclarations } from '../ValidatorShared/FailInterchangeItemRedeclarations';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.namespace.forEach((namespace) => {
    namespace.entity.interchange.forEach((interchange) => {
      if (interchange.elements.length === 0) return;
      failInterchangeItemRedeclarations(
        'InterchangeMustNotRedeclareInterchangeElements',
        'interchange element',
        interchange,
        interchange.elements,
        failures,
      );
    });
  });
  return failures;
}
