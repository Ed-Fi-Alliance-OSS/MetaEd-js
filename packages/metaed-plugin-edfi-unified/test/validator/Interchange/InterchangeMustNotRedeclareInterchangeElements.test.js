// @flow
import { newMetaEdEnvironment, MetaEdTextBuilder, InterchangeBuilder, NamespaceBuilder } from 'metaed-core';
import type { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { validate } from '../../../src/validator/Interchange/InterchangeMustNotRedeclareInterchangeElements';

describe('when validating interchange interchange element has different names', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartInterchange('InterchangeName')
      .withDocumentation('InterchangeDocumentation')
      .withDomainEntityElement('DomainEntityElementName1')
      .withDomainEntityElement('DomainEntityElementName2')
      .withEndInterchange()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new InterchangeBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('edfi');
    failures = validate(metaEd);
  });

  it('should build one interchange', () => {
    expect(coreNamespace.entity.interchange.size).toBe(1);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating interchange interchange element has duplicate names', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityElementName: string = 'DomainEntityElementName';
  let failures: Array<ValidationFailure>;
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartInterchange('InterchangeName')
      .withDocumentation('InterchangeDocumentation')
      .withDomainEntityElement(domainEntityElementName)
      .withDomainEntityElement(domainEntityElementName)
      .withEndInterchange()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new InterchangeBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('edfi');
    failures = validate(metaEd);
  });

  it('should build one interchange', () => {
    expect(coreNamespace.entity.interchange.size).toBe(1);
  });

  it('should have validation failures', () => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('InterchangeMustNotRedeclareInterchangeElements');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot();
    expect(failures[0].sourceMap).toMatchSnapshot();
  });
});

describe('when validating interchange interchange element has multiple duplicate names', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityElementName1: string = 'DomainEntityElementName1';
  const domainEntityElementName2: string = 'DomainEntityElementName2';
  let failures: Array<ValidationFailure>;
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartInterchange('InterchangeName')
      .withDocumentation('InterchangeDocumentation')
      .withDomainEntityElement(domainEntityElementName1)
      .withDomainEntityElement(domainEntityElementName1)
      .withDomainEntityElement(domainEntityElementName2)
      .withDomainEntityElement(domainEntityElementName2)
      .withEndInterchange()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new InterchangeBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('edfi');
    failures = validate(metaEd);
  });

  it('should build one interchange', () => {
    expect(coreNamespace.entity.interchange.size).toBe(1);
  });

  it('should have validation failures', () => {
    expect(failures).toHaveLength(2);
    expect(failures[0].validatorName).toBe('InterchangeMustNotRedeclareInterchangeElements');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot();
    expect(failures[0].sourceMap).toMatchSnapshot();
    expect(failures[1].validatorName).toBe('InterchangeMustNotRedeclareInterchangeElements');
    expect(failures[1].category).toBe('error');
    expect(failures[1].message).toMatchSnapshot();
    expect(failures[1].sourceMap).toMatchSnapshot();
  });
});
