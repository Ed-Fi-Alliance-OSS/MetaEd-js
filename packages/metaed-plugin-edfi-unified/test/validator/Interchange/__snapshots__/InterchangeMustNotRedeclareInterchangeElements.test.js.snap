// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when validating interchange interchange element has duplicate names should have validation failures 1`] = `"Interchange InterchangeName redeclares interchange element DomainEntityElementName."`;

exports[`when validating interchange interchange element has duplicate names should have validation failures 2`] = `
Object {
  "column": 4,
  "line": 5,
  "tokenText": "domain entity",
}
`;

exports[`when validating interchange interchange element has multiple duplicate names should have validation failures 1`] = `"Interchange InterchangeName redeclares interchange element DomainEntityElementName1."`;

exports[`when validating interchange interchange element has multiple duplicate names should have validation failures 2`] = `
Object {
  "column": 4,
  "line": 5,
  "tokenText": "domain entity",
}
`;

exports[`when validating interchange interchange element has multiple duplicate names should have validation failures 3`] = `"Interchange InterchangeName redeclares interchange element DomainEntityElementName2."`;

exports[`when validating interchange interchange element has multiple duplicate names should have validation failures 4`] = `
Object {
  "column": 4,
  "line": 7,
  "tokenText": "domain entity",
}
`;
